sdk/core/AzureCore/Source/Util/ConvertingInitializers.swift

```diff
@@ -57,23 +57,64 @@ extension Date {
             formatter.dateFormat = "EEE, dd MMM yyyy HH:mm:ss z"
             return formatter
         }()
+
+        static var iso8601: DateFormatter = {
+            let formatter = DateFormatter()
+            formatter.calendar = Calendar(identifier: .iso8601)
+            formatter.locale = Locale(identifier: "en_US_POSIX")
+            formatter.timeZone = TimeZone(abbreviation: "GMT")
+            formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZZZZZ"
+            return formatter
+        }()
+
+        static var iso8601WithMilliseconds: DateFormatter = {
+            let formatter = DateFormatter()
+            formatter.calendar = Calendar(identifier: .iso8601)
+            formatter.locale = Locale(identifier: "en_US_POSIX")
+            formatter.timeZone = TimeZone(abbreviation: "GMT")
+            formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSZZZZZ"
+            return formatter
+        }()
     }
 
     public enum Format {
         case rfc1123
+        case iso8601
 
         public var formatter: DateFormatter {
             switch self {
             case .rfc1123:
                 return Formatters.rfc1123
+            case .iso8601:
+                return Formatters.iso8601
+            }
+        }
+
+        fileprivate var alternateFormatters: [DateFormatter] {
```

Comment:
I think exposing `.iso8601(withMilliseconds: true)` would be conceptually simpler.

Reply:
I actually spent a lot of time on this, because it was annoying. In fact, one of the passes I took at it was exactly the suggestion you provided above. However I was trying to reconcile a few things that at first blush don't want to be reconciled:

- The SDK should be opinionated about how dates are formatted as strings
- The SDK must be flexible as to how to parse dates out of strings sent by the server
- The DateFormatter class (and the ISO8601DateFormatter class, which, incredibly, is *NOT* a subclass of DateFormatter) has no ability to be flexible when parsing strings to dates - the string MUST follow the precise format specified in the formatting string.

My attempt to resolve these constraints resulted in the solution you see here. It's asymmetric - the SDK can parse ISO8601 dates in two formats (note: this is still extremely insufficient to be a global solution - the [spec](https://en.wikipedia.org/wiki/ISO_8601) denotes a wide variety of formats that are valid ISO8601 date strings) but has an opinion about the only "correct" format in which a date should be rendered as an ISO8601 date string (and the opinion is that millisecond-accuracy is unnecessary, so we never emit fractional seconds).

Frustratingly, the ISO8601DateFormatter class exists that is aware of all of the different "options" you have available when formatting a date into an ISO8601 date string. But unfortunately it suffers the same limitation as DateFormatter, in that you must set the options to exactly match those present in the date string before parsing it into a Date, otherwise it will return nil even if the string is a valid ISO8601 date string.

I'm totally open to other suggestions here. If there were a way to accept *any* valid ISO8601 date string when parsing a string to a Date, I would much prefer that. And likewise if it could be encoded into the API elegantly, I would love to allow an option set of `ISO8601DateFormatter.Options` to be provided when formatting a Date to a string in ISO8601 format.

Reply:
I decided to remove the fallback logic for now. iOS's built-in `JSONDecoder.DateDecodingStrategy` exposes `iso8601` as a strategy which also only parses dates without milliseconds so I feel like there is at least some prior art. And right now I'm only using it for parsing the date in the SAS token which lacks the milliseconds part anyhow. If/when we need to support iso8601 dates with milliseconds or other less common formatting options we can revisit.